crates/ruff_python_semantic/src/analyze/typing.rs

```diff
@@ -382,6 +382,22 @@ pub fn is_mutable_expr(expr: &Expr, semantic: &SemanticModel) -> bool {
 pub fn is_type_checking_block(stmt: &ast::StmtIf, semantic: &SemanticModel) -> bool {
     let ast::StmtIf { test, .. } = stmt;
 
+    if semantic.use_new_type_checking_block_detection_semantics() {
```

Comment:
I think this version is now slightly different to what you had before as it removes support for `if False:` and `if 0:`.